File: packages/adapter/src/walk.ts
/**
 * The one value-tree walker behind `maskSecretsWith` and `maskLogValueWith`.
 * It masks everything JSON serialization would emit, so no shape of value
 * reaches a host's serializer unmasked.
 */

import { CYCLE_MARKER, DEFAULT_LIMITS, ERROR_MARKER, LIMIT_MARKER, maskLeafWith, resolveLimit } from "./mask-leaf.js";
import type { Limits, MaskOptions, Policy, ScanAndRedact } from "./types.js";

export interface WalkContext {
  readonly policy: Policy;
  readonly limits: Limits;
  readonly budget: { leaves: number };
}

/** Per-key fallback to `DEFAULT_LIMITS`, so `{ maxDepth: undefined }` or `NaN` never disables a bound. */
export function resolveLimits(overrides: Partial<Limits> | undefined): Limits {
  const limits: { -readonly [K in keyof Limits]: number } = { ...DEFAULT_LIMITS };
  for (const key of Object.keys(DEFAULT_LIMITS) as (keyof Limits)[]) {
    limits[key] = resolveLimit(overrides?.[key], DEFAULT_LIMITS[key]);
  }
  return limits;
}

export function createWalkContext(options: MaskOptions): WalkContext {
  const limits = resolveLimits(options.limits);
  return { policy: options.policy, limits, budget: { leaves: limits.maxTotalLeaves } };
}

export function isPlainObject(value: unknown): value is Record<string, unknown> {
  if (typeof value !== "object" || value === null) return false;
  const proto = Object.getPrototypeOf(value);
  return proto === Object.prototype || proto === null;
}

export function maskString(scanAndRedact: ScanAndRedact, value: string, ctx: WalkContext): string {
  if (ctx.budget.leaves <= 0) return LIMIT_MARKER;
  ctx.budget.leaves -= 1;
  return maskLeafWith(scanAndRedact, value, {
    policy: ctx.policy,
    maxStringLength: ctx.limits.maxStringLength,
  });
}

/** Defines a data key without invoking inherited setters such as `__proto__`. */
export function defineDataKey(out: object, key: string, value: unknown): void {
  Object.defineProperty(out, key, {
    value,
    enumerable: true,
    configurable: true,
    writable: true,
  });
}

type Walk = (value: unknown, depth: number) => unknown;

/**
 * Copies `source`'s own enumerable string keys into `out`, masked. Keys past
 * `maxObjectKeys` are dropped, never passed through; a throwing getter
 * becomes {@link ERROR_MARKER} for that key alone.
 */
function maskProperties(
  source: object,
  out: object,
  walk: Walk,
  ctx: WalkContext,
  depth: number,
  skip?: ReadonlySet<string>,
): void {
  const record = source as Record<string, unknown>;
  const keys = Object.keys(source).filter((key) => !skip?.has(key));
  for (const key of keys.slice(0, ctx.limits.maxObjectKeys)) {
    let masked: unknown;
    try {
      masked = walk(record[key], depth + 1);
    } catch {
      masked = ERROR_MARKER;
    }
    defineDataKey(out, key, masked);
  }
}

// Own properties that would clash with the fixed fields are skipped; `cause`
// is walked once, below, whether or not it is enumerable.
const ERROR_OWN_KEYS: ReadonlySet<string> = new Set(["type", "message", "stack", "cause"]);

/**
 * An `Error` becomes `{ type, message, stack, ...ownProps, cause }`, the
 * shape `pino.stdSerializers.err` emits. `message`, `stack` and `cause` are
 * usually non-enumerable, so a plain property walk would drop them — or,
 * for a host serializer that reads them itself, leave them unmasked.
 */
function maskError(error: Error, scanAndRedact: ScanAndRedact, walk: Walk, ctx: WalkContext, depth: number) {
  const out: Record<string, unknown> = {};
  const name: unknown = error.name;
  out.type = typeof name === "string" ? maskString(scanAndRedact, name, ctx) : "Error";
  out.message = maskString(scanAndRedact, String(error.message ?? ""), ctx);
  if (typeof error.stack === "string") out.stack = maskString(scanAndRedact, error.stack, ctx);
  maskProperties(error, out, walk, ctx, depth, ERROR_OWN_KEYS);
  if (error.cause !== undefined) out.cause = walk(error.cause, depth + 1);
  return out;
}

function maskObject(value: object, scanAndRedact: ScanAndRedact, walk: Walk, ctx: WalkContext, depth: number) {
  if (value instanceof Error) return maskError(value, scanAndRedact, walk, ctx, depth);
  if (Array.isArray(value)) {
    // Elements beyond the limit are dropped, never passed through unmasked.
    return value.slice(0, ctx.limits.maxArrayLength).map((item) => walk(item, depth + 1));
  }
  // Serialize the way JSON.stringify would: a boxed primitive is its
  // primitive, and `toJSON()` replaces the value (Date, URL, Buffer, ...).
  // Checked on plain objects too: a toJSON left on the masked copy would
  // run again at serialization time and emit its unmasked result.
  if (value instanceof String) return maskString(scanAndRedact, value.valueOf(), ctx);
  if (value instanceof Number || value instanceof Boolean) return value.valueOf();
  const toJSON: unknown = (value as { toJSON?: unknown }).toJSON;
  if (typeof toJSON === "function") {
    let json: unknown;
    try {
      json = toJSON.call(value, "");
    } catch {
      return ERROR_MARKER;
    }
    // One level deeper, so a toJSON that returns a fresh toJSON object
    // each call still terminates at maxDepth.
    return walk(json, depth + 1);
  }
  // Plain objects, and any other instance (class, IncomingMessage, Map, ...)
  // as the own enumerable properties JSON.stringify would emit.
  const out = {};
  maskProperties(value, out, walk, ctx, depth);
  return out;
}

/**
 * Masks every string reachable in `data`. Never throws: a value that cannot
 * be read (throwing getter or `toJSON`, revoked proxy) becomes
 * {@link ERROR_MARKER}, a repeat visit {@link CYCLE_MARKER}, and anything
 * past a budget {@link LIMIT_MARKER}. Numbers, booleans, `null`,
 * `undefined`, bigints, symbols and functions pass through unchanged.
 */
export function walkValue(scanAndRedact: ScanAndRedact, data: unknown, options: MaskOptions): unknown {
  const ctx = createWalkContext(options);
  const seen = new Set<object>();
  const walk: Walk = (value, depth) => {
    if (typeof value === "string") return maskString(scanAndRedact, value, ctx);
    if (typeof value !== "object" || value === null) return value;
    if (depth >= ctx.limits.maxDepth) return LIMIT_MARKER;
    if (seen.has(value)) return CYCLE_MARKER;
    seen.add(value);
    try {
      return maskObject(value, scanAndRedact, walk, ctx, depth);
    } catch {
      return ERROR_MARKER;
    } finally {
      seen.delete(value);
    }
  };
  return walk(data, 0);
}

File: python/redact_secret_adapters/_walk.py
"""The one value-tree walker (L2) behind ``mask_secrets_with`` and
``mask_log_value_with``, mirroring ``packages/adapter/src/walk.ts`` and the
walk in ``mask-log-value.ts``.

Walked: ``str`` leaves; ``dict`` and its subclasses (``OrderedDict``,
``defaultdict``, ...), returned as a plain ``dict``; ``list`` and its
subclasses, returned as a plain ``list``; tuples, returned as a plain
``tuple``; and exceptions, returned as a ``{"type", "message", "stack",
"cause"}`` mapping. Everything else is returned unchanged.
"""

from __future__ import annotations

import traceback
from typing import Any, Callable, Optional

from .mask_leaf import CYCLE_MARKER, DEFAULT_LIMITS, ERROR_MARKER, LIMIT_MARKER, mask_leaf_with


class _Walk:
    __slots__ = ("scan_and_redact", "policy", "limits", "leaves", "seen")

    def __init__(self, scan_and_redact: Callable[..., Any], policy: Any, limits: Optional[dict[str, int]]) -> None:
        self.scan_and_redact = scan_and_redact
        self.policy = policy
        self.limits = {**DEFAULT_LIMITS, **(limits or {})}
        self.leaves = self.limits["max_total_leaves"]
        self.seen: set[int] = set()

    def string(self, value: str) -> str:
        if self.leaves <= 0:
            return LIMIT_MARKER
        self.leaves -= 1
        return mask_leaf_with(
            self.scan_and_redact, value, policy=self.policy, max_string_length=self.limits["max_string_length"]
        )

    def stack(self, exc: BaseException) -> str:
        try:
            formatted = "".join(traceback.format_exception(type(exc), exc, exc.__traceback__))
        except Exception:
            return ERROR_MARKER
        return self.string(formatted)

    def exception(self, exc: BaseException, depth: int) -> dict[str, Any]:
        try:
            message = str(exc)
        except Exception:
            # A raising __str__ must not escape into logger.exception().
            message = None
        out: dict[str, Any] = {
            "type": type(exc).__name__,
            "message": ERROR_MARKER if message is None else self.string(message),
            "stack": self.stack(exc),
        }
        if exc.__cause__ is not None:
            out["cause"] = self.value(exc.__cause__, depth + 1)
        return out

    def value(self, value: Any, depth: int) -> Any:
        if isinstance(value, str):
            return self.string(value)
        if not isinstance(value, (BaseException, list, tuple, dict)):
            # Numbers, booleans, None, and other objects (sets, dataclasses,
            # datetimes, ...) are returned unchanged.
            return value

        if depth >= self.limits["max_depth"]:
            return LIMIT_MARKER
        if id(value) in self.seen:
            return CYCLE_MARKER
        self.seen.add(id(value))
        try:
            if isinstance(value, BaseException):
                return self.exception(value, depth)
            if isinstance(value, dict):
                # Keys beyond the limit are dropped, never passed through unmasked.
                keys = list(value.keys())[: self.limits["max_object_keys"]]
                return {key: self.value(value[key], depth + 1) for key in keys}
            # Elements beyond the limit are dropped, never passed through unmasked.
            masked = [self.value(item, depth + 1) for item in value[: self.limits["max_array_length"]]]
            return tuple(masked) if isinstance(value, tuple) else masked
        finally:
            self.seen.discard(id(value))


def walk(
    scan_and_redact: Callable[..., Any], data: Any, *, policy: Optional[Any], limits: Optional[dict[str, int]]
) -> Any:
    return _Walk(scan_and_redact, policy, limits).value(data, 0)


def mask_exception_text_with(
    scan_and_redact: Callable[..., Any],
    exc: BaseException,
    *,
    policy: Optional[Any] = None,
    limits: Optional[dict[str, int]] = None,
) -> str:
    """The masked ``stack`` that walking ``exc`` would produce, without
    scanning the message or cause separately: the formatted traceback
    already contains both, and the logging filter keeps only this text."""
    state = _Walk(scan_and_redact, policy, limits)
    if state.limits["max_depth"] <= 0:
        return LIMIT_MARKER
    return state.stack(exc)
